### ISSUE TITLE
Asynchronous functions in plugins not supported

### ISSUE THREAD

**Issue by saurabhdaware** [abelljs/abell#36]
**Describe the bug**
There is no way to make `beforeBuild` function from plugins, asynchronous. It is a common use-case that someone would want to fetch data in the plugin before building the files.

**Expected behavior**

Plugin with this code should work
```js

async beforeBuild(programInfo) {
  await fetch('example.com');
  // write into file and other operations.
}
module.exports = { beforeBuild };

```

**Comment by saurabhdaware** [comment]
This is the line that executes beforeBuild function https://github.com/abelljs/abell/blob/main/src/build.js#L50

**Comment by anuraghazra** [comment]
So i yesterday i implemented this with a plugin: https://github.com/anuraghazra/abell-issues-blog

As you can see in the workflow file i had to do "npm run build" twice in order to fetch the contents from remote api first... the again for making sure abell spits out generated HTML files.

Now as far as i can tell by looking at the code of abell i think fundamentally plugin system is have some issues

In the cli.js i evaluated the code and found out that

https://github.com/abelljs/abell/blob/61c8d174b1abf273f1b8a89856a94272b3f41cfc/src/cli.js#L10-L22

In line 13 the code is function "getBaseProgramInfo" is called which i think gathers all the information about files including the "contentDirectories" array. the getBaseProgramInfo calls the "loadContent" function here:

https://github.com/abelljs/abell/blob/61c8d174b1abf273f1b8a89856a94272b3f41cfc/src/utils/build-utils.js#L100-L103 

which recursively finds all the `recursiveFindFiles(contentPath, '.md')` .md files and returns the content directories array.


So the reason behind abell plugins are failing to await the response for the build is because you can see that in `cli.js` line 13 the programInfo is first called and at that moment it will not have any contentDirectories because the "beforeBuild" & "afterBuild" plugins are called inside the build() function.... thats why when calling "npm run build" again contents folders will have all the remotely fetched contents inside and the `getBaseProgramInfo` can detect the `contentDirectories`.

This is the reason i think we had to do "npm run abell" twice in order to force abell fetch the content on the first run & then generate the html and the second run.

**Comment by saurabhdaware** [comment]
No actually contentDirectories will be there synchronously available in `beforeBuild`. So in `cli.js` getBaseProgramInfo will gather all information including `contentDirectories`.

This problem does not really have anything to do with contentDirectories and baseInfo.

The reason is, you have await in fetch API of plugin and the function is async, but Abell doesn't wait for the info from API to be fetched and next part to be executed, it will execute the build anyway and after building files, it will run the plugin.

For it to work synchronously, we need to await on https://github.com/abelljs/abell/blob/main/src/commands/build.js#L46

**Comment by anuraghazra** [comment]
"No actually contentDirectories will be there synchronously available in beforeBuild. So in cli.js getBaseProgramInfo will gather all information including contentDirectories." actually it is the contentDirectories on the first run of the abell build content directories are empty. and because  while the api is fetching the contents the the getBaseProgramInfo is already done.


And i actually the issue is solved... you have to extract the beforeBuild plugins on cli.js file so that you can execute the beforeBuild plugin before the getBaseProgramInfo.

**Comment by saurabhdaware** [comment]
> Actually it is the contentDirectories on the first run of the abell build content directories are empty. and because while the api is fetching the contents the the getBaseProgramInfo is already done.

Oh ok got it! even in that case, we will need to await right?

**Comment by anuraghazra** [comment]
yes of course we need to await. but on the program.command('build').action()

**Comment by saurabhdaware** [comment]
gotcha! In that case we won't be able to send `programInfo` to beforeBuild, do you think it is needed?

**Comment by saurabhdaware** [comment]
how about keep it as it is right now but let user's return new `programInfo` object where they can return modified values? in that case we can return new contentArray.

Which will be better solution?

**Comment by anuraghazra** [comment]
Yes we could send the programInfo.... because we will be needing the programInfo to get the abellConfig.plugins (i know you also have a getAbellConfig function that would also work).

Now its up to you if you wanted to pass the programInfo to the beforeBuild hook or not.

**Comment by saurabhdaware** [comment]
Yes users will need `programInfo`, maybe we can separate content gathering logic from getBaseProgramInfo function, then 
1. run getBaseProgramInfo and generate programInfo without content related values
2. await beforeBuild plugin and pass that programInfo
3. Gather content and add to programInfo.

**Comment by anuraghazra** [comment]
I actually did the exact same thing... i initially tried to do it with the just calling "loadContent" function and for some reason that throws error... i think you need to separate the logic

**Comment by anuraghazra** [comment]
But currently what i did is just called the getBaseProgramInfo twice first before the plugin and then after the plugin

**Comment by saurabhdaware** [comment]
Resolved in #44 #40
